Lesson09

lunarnuts opened the pull request.

lunarnuts committed:
init

gordiyanovp commented on lesson09/go.sum:
```diff
@@ -0,0 +1,2 @@
+github.com/lib/pq v1.10.2 h1:AqzbZs4ZoCBp+GtejcpCpcxM3zlSMx29dXbUSeVtJb8=
```
I have state several times, that `go.mod` file needs to be in your package, not just `go.sum`. Modify your `.gitignore` and push it as well. this is also not fixed in lesson07 either.

gordiyanovp commented on lesson09/src/cmd/migrate/migrate.go:
```diff
@@ -0,0 +1,74 @@
+package main
```
Did you just copy an example from `github.com/wshaman/migrate`? I would expect the migration integrated into your service: when it is started it performs migration from an `.sql` file, containing database schema, because at the moment there is no way to know what the database structure is supposed to be.

gordiyanovp commented on lesson09/src/cmd/api/api.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func main() {
+	dbObj, err := db.New(dbsFromEnv())
```
The hw description says to have the following functionality:
- Add contact
- Assign contact to a group
- List ALL contacts by group

Right now there is only an example of usage of the last one.

gordiyanovp reviewed (commented):
Right now it is hard to test how this as there is no way for me to set up a database. I assume I can start a postgres database locally, but even then I know nothing about it's schema as the migration is not set up properly. 
Lack of `go.mod` also makes it hard to launch the app.

lunarnuts committed:
upd

lunarnuts committed:
upd

lunarnuts committed:
upd2

gordiyanovp commented on lesson09/src/db/migrations/init.up.sql:
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE groups (
```
This is not what is meant by migrations. Migration needs to be applied at the application start. You had a library here, but it was just a separate app, that received it's own requests. You may use that one, but need to integrate it into the main application, and perform migration for the database, based on the predefined setup

lunarnuts replied:
added migration using migrate from wshaman/migrate pkg, refactored, created migration schemas and integrated into main function.

lunarnuts committed:
add migrate

gordiyanovp commented on lesson09/src/cmd/api/main.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"github.com/gorilla/mux"
 	_ "github.com/lib/pq"
 	"github.com/lunarnuts/go-course/tree/lesson09/src/cmd/api/handlers"
+	"github.com/lunarnuts/go-course/tree/lesson09/src/cmd/migrate"
```
This package does not exist

lunarnuts replied:
sorry, forgot to commit it.

lunarnuts committed:
pkg migrate

gordiyanovp commented on lesson09/src/cmd/migrate/migrate.go:
```diff
@@ -0,0 +1,39 @@
+	_ "github.com/lib/pq"
+
+	"github.com/lunarnuts/go-course/tree/lesson09/src/db"
+	_ "github.com/lunarnuts/go-course/tree/lesson09/src/db/migrations"
```
This won't compile as the directory does not contain any go files.
Still, even when removing this line, the migration is not performed

lunarnuts replied:
I dont understand why, but git add . did not add all files.

gordiyanovp reviewed (commented):
Migration

lunarnuts committed:
upload migrations

gordiyanovp commented on lesson09/src/cmd/api/main.go:
```diff
@@ -0,0 +1,82 @@
+		func(w http.ResponseWriter, r *http.Request) {
+			conn, _ := db.AcquireConn(pool)
+			handlers.AssignContactToGroup(&conn, w, r)
+		}).Name("AssignContactsToGroup").Methods(http.MethodGet)
```
You probably want to make this POST and accept JSON, as it seems more RESTful to have POST on modifying or adding a record.

gordiyanovp reviewed (commented):
From what I can see, your set of endpoints doesn't allow to create a group and trying to add any new contact fails because of FOREIGN KEY constraints. There are no groups and adding a new contact requires a group ID. You can either create a set of groups during a migration or make an endpoint, which allows you to create a custom group.

lunarnuts committed:
updated models, handlers  and main.go

lunarnuts commented:
> From what I can see, your set of endpoints doesn't allow to create a group and trying to add any new contact fails because of FOREIGN KEY constraints. There are no groups and adding a new contact requires a group ID. You can either create a set of groups during a migration or make an endpoint, which allows you to create a custom group.

Well i did create  groups during migration. The error was in my models pkg, i made an error in Select query to db, therefore assignment to group did not work. Now i changed it to receive json and changed method to put.
